ocis-reva/docs/storages.md

```diff
@@ -0,0 +1,194 @@
+---
+title: "Storages"
+date: 2020-04-27T18:46:00+01:00
+weight: 37
+geekdocRepo: https://github.com/owncloud/ocis-reva
+geekdocEditPath: edit/master/docs
+geekdocFilePath: storages.md
+---
+
+## Storage commands
+
+`ocis-reva` has multiple storage provider commands to preconfigure different default configurations for the reva *storage provider* service. While you could rerun `ocis-reva storage-oc` multiple times with different flags to get multiple instances we are giving the different commands the necessary default configuration to allow the `ocis` binary to simply start them and not deal with configuration.
+
+## Storage providers
+
+To manage the file tree ocis uses reva *storage providers* that are accessing the underlying storage using a *storage driver*. The driver can be used to change the implementation of a storage aspect to better reflect the actual underlying storage capabilities. As an example a move operation on a POSIX filesystem ([theoretically](https://danluu.com/deconstruct-files/)) is an atomic operation. When trying to implement a file tree on top of S3 there is no native move operation that can be used. A naive implementation might fall back on a COPY and DELETE. Some S3 implementations provide a COPY operation that uses an existing key as the source, so the file at least does not need to be reuploaded. In the worst case scenario, which is renaming a folder with hundreds of thousands of objects, a reupload for every file has to be made. Instead of hiding this complexity a better choice might be to disable renaming of files or at least folders on S3. There are however implementations of filesystems on top of S3 that store the tree metadata in dedicated objects or use a completely different persistence mechanism like a distributed key value store to implement the file tree aspect of a storage.
+
+
+{{< hint info >}}
+While the *storage provider* is responsible for managing the tree, file up and download is delegated to a dedicated *data provider*. See below.
+{{< /hint >}}
+
+## Storage aspects
```

Comment:
Codacy found an issue: [[heading-style] Headings should use setext](https://app.codacy.com/gh/owncloud/ocis/pullRequest?prid=6202733)